### PR TITLE
feat(NODE-3784):  Add `enableUtf8Validation` option

### PR DESCRIPTION
**Pull request by baileympearson**
#### What is changing?

A new option was added to the db, client, operation and collection classes to allow users to enable/disable utf8 validation.

The legacy response class is not modified - deserialization changes are only present in the `BinMsg` class.

##### Is there new documentation needed for these changes?

Yes

#### What is the motivation for this change?

UTF8 validation is now enabled by default.  Disabling utf8 validation can result in a performance benefit for users.

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

File: src/cmap/commands.ts [hunk 1]
```diff
@@ -876,4 +876,15 @@ export class BinMsg {
 
     this.parsed = true;
   }
+
+  private _parseBsonSerializationOptions({ enableUtf8Validation }: BSONSerializeOptions): {
+    utf8: { writeErrors: false } | false;
+  } {
+    // eslint-disable-next-line no-restricted-syntax
+    if (enableUtf8Validation === undefined || enableUtf8Validation) {
```
Comment (baileympearson) [inline src/cmap/commands.ts:884]: I _think_ this should never be undefined, because I set the default value to `true` in both `connection_string.ts` and `resolveBSONOptions`.  But the interface does allow for an undefined, so I decided to handle it.  Checking for `undefined` results in a linting error so I have to use an `eslint-disable`

File: src/cmap/commands.ts [hunk 2]
```diff
@@ -876,4 +875,14 @@ export class BinMsg {
 
     this.parsed = true;
   }
+
+  private _parseBsonSerializationOptions({ enableUtf8Validation }: BSONSerializeOptions): {
```
Comment (nbbeeken) [inline src/cmap/commands.ts:879]:
```suggestion
  parseBsonSerializationOptions({ enableUtf8Validation }: BSONSerializeOptions): {
```

Since we're working on an internal class here, everything's inherently private, its stylistic but the access modifier likely would just get in the way of unit tests written in typescript if that ever came about.

File: src/cmap/commands.ts [hunk 3]
```diff
@@ -856,7 +855,7 @@ export class BinMsg {
       if (payloadType === 0) {
         const bsonSize = this.data.readUInt32LE(this.index);
         const bin = this.data.slice(this.index, this.index + bsonSize);
-        this.documents.push(raw ? bin : BSON.deserialize(bin, bsonOptions));
+        this.documents.push(raw ? bin : deserialize(bin, bsonOptions));
```
Comment (nbbeeken) [inline src/cmap/commands.ts:858]: Is this change (`BSON.des` vs just `des`) something to make sinon work? I think that's fine let's just make it the same for all usages and serialize usages.

File: src/cmap/commands.ts [hunk 4]
```diff
@@ -6,6 +6,7 @@ import type { Long, Document, BSONSerializeOptions } from '../bson';
 import type { ClientSession } from '../sessions';
 import type { CommandOptions } from './connection';
 import { MongoRuntimeError, MongoInvalidArgumentError } from '../error';
+import { deserialize } from 'bson';
```
Comment (nbbeeken) [inline src/cmap/commands.ts:9]: You have to import BSON things from `src/bson.ts` to continue supporting our optional dependency logic.

File: test/functional/cmap/commands.test.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+            { name: 'John Doe' },
+            passOptionTo === 'operation' ? option : {}
+          );
+
+          expect(deserializeSpy.called).to.be.true;
+          const validationArgument = deserializeSpy.lastCall.lastArg.validation;
+          expect(validationArgument).to.deep.equal(EXPECTED_VALIDATION_DISABLED_ARGUMENT);
+        } finally {
+          await client.close();
+        }
```
Comment (nbbeeken) [inline test/functional/cmap/commands.test.ts:51]: Nothing to change here, just calling out that normally we would like to use  an `afterEach` hook to ensure the client closes when each test finishes, unfortunately our session leak checker is currently implemented as an afterEach hook that runs before the one you would define here. So your hook wouldn't get the chance to clean up the sessions/client before the leak checker would throw.

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: src/cmap/commands.ts (hunk 2 above)
Comment (baileympearson) [reply src/cmap/commands.ts]: Fixed!

File: src/cmap/commands.ts (hunk 4 above)
Comment (baileympearson) [reply src/cmap/commands.ts]: ok yeah good catch.  This was to make Sinon work, but I had to do this because I wasn't importing from `src/bson`.

File: src/cmap/commands.ts [hunk 6]
```diff
@@ -6,6 +6,7 @@ import type { Long, Document, BSONSerializeOptions } from '../bson';
 import type { ClientSession } from '../sessions';
 import type { CommandOptions } from './connection';
 import { MongoRuntimeError, MongoInvalidArgumentError } from '../error';
+// import { deserialize } from '../bson';
```
Comment (durran) [inline src/cmap/commands.ts:9]: I'd say after the revert we can just remove this now.

**Review by durran — CHANGES_REQUESTED** [review]

File: src/cmap/commands.ts (hunk 6 above)
Comment (baileympearson) [reply src/cmap/commands.ts]: Oh, good catch!

File: src/cmap/commands.ts (hunk 3 above)
Comment (baileympearson) [reply src/cmap/commands.ts]: Yup, this was just to sinon work.  But importing from `src/bson` instead of `'bson'` fixed the issue that I was seeing earlier, and I can revert it.

**Review by nbbeeken — APPROVED** [review]

File: src/bson.ts [hunk 7]
```diff
@@ -55,6 +55,9 @@ export interface BSONSerializeOptions
     > {
   /** Return BSON filled buffers from operations */
   raw?: boolean;
+
+  /** Enable utf8 validation when deserializing BSON documents.  Defaults to true. */
+  enableUtf8Validation?: boolean;
```
Comment (dariakp) [inline src/bson.ts:60]: Leaving this comment on the first file for visibility: let's wait to merge this PR until we decide whether or not we're going to do another patch release before the holidays

**Review by dariakp — CHANGES_REQUESTED** [review]

**Review by durran — DISMISSED** [review]

File: test/unit/commands.test.ts [hunk 8]
```diff
@@ -37,45 +38,64 @@ const invalidUtf8InWriteErrorsJSON = {
...
-      errmsg:
-        'E11000 duplicate key error collection: bigdata.test index: text_1 dup key: { text: "d☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃�..." }'
-    }
-  ],
-  ok: 1
-};
 
 describe('BinMsg BSON utf8 validation', () => {
-  context('when validation is disabled for writeErrors', () => {
+  test('invalid utf8 data throws as expected when deserialized', () => {
```
[the remark this answers is not in the document]
Comment (baileympearson) [reply test/unit/commands.test.ts:44]: This is a poorly written test description, I'll update it (whether or not we keep the test is another question).

Grace and @nbbeeken wrote this test to confirm that the mock data they manually created for these unit tests was correctly malformed (it did contain invalid utf8).  I think Neal wanted to keep this, so I left it.  Do you have a preference whether or not we keep it?

File: test/unit/commands.test.ts [hunk 9]
```diff
@@ -37,45 +38,64 @@ const invalidUtf8InWriteErrorsJSON = {
...
-      errmsg:
-        'E11000 duplicate key error collection: bigdata.test index: text_1 dup key: { text: "d☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃☃�..." }'
-    }
-  ],
-  ok: 1
-};
 
 describe('BinMsg BSON utf8 validation', () => {
-  context('when validation is disabled for writeErrors', () => {
+  test('mock data contains replacement characters for invalid utf8 in writeError object', () => {
```
Comment (dariakp) [inline test/unit/commands.test.ts:44]:
```suggestion
  test('bson correctly deserializes data with replacement characters for invalid utf8 in writeErrors object', () => {
  // this is a sanity check to make sure nothing unexpected is happening in the deserialize method itself
```

**Review by dariakp — CHANGES_REQUESTED** [review]
Just a small suggestion for phrasing on the test case

File: test/unit/commands.test.ts (hunk 9 above)
Comment (baileympearson) [reply test/unit/commands.test.ts]: Done!

**Review by dariakp — APPROVED** [review]

**Review by durran — APPROVED** [review]
